src/Streaming/Internal/Many.hs

```diff
@@ -98,6 +108,163 @@ unzip = loop
 {-# INLINABLE unzip #-}
 
 
+{- Remarks on the design of zip functions
+
+Zip functions have two design choices:
+(1) What do we do with the end-of-stream values of both streams?
+(2) If the streams are of different length, do we keep or throw out the
+remainder of the longer stream?
+
+* We are assuming not to take infinite streams as input and instead deal with
+reasonably small finite streams.
+* To avoid making choices for the user, we keep both end-of-stream payloads
+* The default zips (ones without a prime in the name) use @effects@ to consume
+the remainder stream after zipping. We include zip function variants that
+return no remainder (for equal length streams), or the remainder of the
+longer stream.
+
+-}
+
+-- | The remainder of zipping two streams
+type ZipResidual a b m r1 r2 =
+  ( Either r1 (Stream (Of a) m r1)
+  , Either r2 (Stream (Of b) m r2)
+  )
+
+-- | @zipWith'@ zips two streams applying a function along the way,
+-- keeping the remainder of zipping if there is one.
+zipWith' :: Control.Monad m =>
```

Comment:
Let's have a better naming convention for these. `zipWithR` maybe? For “remainder”. Or something to this effect.